### ISSUE TITLE
Texture (filepath, etc.) Aliases

### ISSUE THREAD

**Issue by Python-Sargent** [minetest/minetest#15535]
### Problem

This may be a rather unusual problem, but I've run into an issue where I have a large amount of textures that are named a certain way, and for compatibility reasons I don't want to/can't change the names of each of those textures.

### Solutions

What I am suggesting is to add the functionality to have aliases for textures and other media, for example I could register an alias for "default_dirt.png", and another mod's dirt texture, whereby it acts like a texturepack, without creating a whole new texture or modifying the existing node. One of the specific usages for this is when you have optional/limited availability media that you want to make plug-and-play.

### Alternatives

Another option is to have the ability to specify a list of textures that are chosen following a set of set rules. For example you could have the base texture be overriden by another texture if it has the overrides_others flag or a similar functionality.

### Additional context

The alternative would have the upperhand of being able to use different textures for different things, for example I've been helping with the game 1042 and the textures as you may know are a bit peculiar, if the alternative were implemented it cold be made so that the setup could be kept the same and there could be an alternative texturepack, which would not break that game when it was not used. The only problem with the alternative solution is it involves overriding nodes which I previously stated was something I hoped to avoid, although I think a mixture of these two solutions would fix the issue with ease.

**Comment by Python-Sargent** [comment]
Sorry for the text wall, I hoped to get my point across better, but you can't expect a writer to not use a lot of words.

**Comment by nininik0** [comment]
YES! I was about to make a request for this, like texture aliases lol maybe I have entities using a certain texture and I've renamed it and I deleted the other one, and some are left textureless which would greatly help it by allowing such festures.

**Comment by nininik0** [comment]
YES! I was about to make a request for this, like texture aliases lol maybe I have entities using a certain texture and I've renamed it and I deleted the other one, and some are left textureless which would greatly help it by allowing such features.

**Comment by appgurueu** [comment]
Related: #3370 (closed due to lack of core dev support), #12694 (might be implemented eventually, I can see the motivation here; would to an extent alleviate this problem by allowing texture pack authors to deal with texture renames better).

That said: I'm not quite sure I understand the problem you're trying to solve or rather, why you want the engine to solve it in a likely inflexible way.

> for compatibility reasons I don't want to/can't change the names of each of those textures

Genuine question: Then why do it? Just leave the names as-is.

If that is not an option, follow-up: Why not define variables / "aliases" in your own code; why does the engine need to do this for you?

> for example I could register an alias for "default_dirt.png", and another mod's dirt texture, whereby it acts like a texturepack, without creating a whole new texture or modifying the existing node

Just reuse the other mod's dirt texture in your mod; no need to have an engine "alias" feature here?

> Another option is to have the ability to specify a list of textures that are chosen following a set of set rules. For example you could have the base texture be overriden by another texture if it has the overrides_others flag or a similar functionality.

This is too vague. How / when is this flag set? At runtime? At load time? Why can't you just register the node with the correct textures to begin with?

**Comment by Python-Sargent** [comment]
> Why not define variables / "aliases" in your own code; why does the engine need to do this for you?
The reason, is because the filename is the only thing provided to the engine, the engine then has to go and search for that texture and apply it. What I want is the ability to have a texture which if present will overwrite the aliased texture. Because the engine handles finding the textures, making the same functionality would involve my code replacing the original texture definition with my own, and I would have to check to see if the texture did exist.

> Just reuse the other mod's dirt texture in your mod; no need to have an engine "alias" feature here?
Like I said, I'm handling textures that are optional/limited availability so I cannot just 'use the texture in my mod'

> This is too vague. How / when is this flag set? At runtime? At load time? Why can't you just register the node with the correct textures to begin with?

The flag would be set during the definition for the node, or for example during the redefinition. As I said this does involve modifying the original definition of the node.

**Comment by appgurueu** [comment]
> What I want is the ability to have a texture which if present will over[ride] the aliased texture. 

Texture overrides (using the same name) are a feature as-is. Why is that not sufficient?

> Like I said, I'm handling textures that are optional/limited availability so I cannot just 'use the texture in my mod'

What does "optional / limited availability" mean here? Can you give a concrete example? So far I understand: You have another mod, which brings its own texture. Conditionally (?) you want to override that mod's textures, at load time (I would argue that this is already somewhat dirty and not something the engine need encourage), but a texture pack doesn't fit the bill. So what stops you from just overriding that mod's textures in a mod, and if you need to use the same textures, referring to them by that mod's texture names?

> As I said this does involve modifying the original definition of the node.

Then I don't see why you can't just directly override the textures?

---

Put very simply: All texture name aliases would do is to let you call `a.png` by a different name, say `b.png`. But I don't see how that solves any problem.

If you're a game developer and you want to rename a texture, you can just do that. If you want to maintain texture pack compatibility, don't rename it. You can still add your own abstractions in your own code to refer to the texture via an alias. The same goes for mods you fully control more generally.

Now okay, suppose you have another mod which you don't control. Doesn't matter, you just have to use the names that mod establishes. You can still override media. You just have to stick to the same names. Inside your own mod, you can still have, say, a simple lookup table of aliases to use another name in your code instead, if you prefer that.

Contrast this with #12694, which is well motivated: Texture packs do not control the games to which they are applied; texture packs might want to use the same textures for different mods and games. Hence they could use a mechanism for specifying aliases to avoid the need to duplicate / symlink files. (This is also easier to implement, because texture packs are on the client, and the client does the "looking up textures by name" thing.)
